### DISCUSSION TITLE
Duplicate Captures

### DISCUSSION THREAD

**Discussion by MixusMinimax** [tree-sitter/tree-sitter#2586] [Q&A] [answered]
Hi! I have already read through the other posts about this, and none of them seem to be like my case. I am using web-tree-sitter, and it gives me different results from what the rust CLI gives me during testing.

### Source string:

```
module core::events;
```

### Parse tree:

```
(source_file [0, 0] - [0, 20]
  (module [0, 0] - [0, 19]
    path: (path [0, 7] - [0, 19]
      (identifier [0, 7] - [0, 11])
      (identifier [0, 13] - [0, 19]))))
```

### Query:

```
(source_file
  (module
    path: (path
      [
        (identifier) @namespace
        _
      ]*)))
```

### The result using the rust cli:

```
examples/example-file
  pattern: 0
    capture: 0 - parent_path, start: (0, 7), end: (0, 11), text: `core`
    capture: 0 - parent_path, start: (0, 13), end: (0, 19), text: `events`
```

### Result using web-tree-sitter:

```js
queries.module.captures(tree.rootNode).map(s => s.node.text)
=> ['core', 'core', 'core', 'events', 'events', 'events']

// also, the node ids and capture names are the same within each group of three.
```

However, if I use `.matches`, it gives me one match with two captures, which is what I would expect.

Could this be a problem with web-tree-sitter or am I doing something wrong? All the other posts on here were missing an anchor or something, but I don't think this is the case here.

When using a quantifier (like `*` or `+`), that expression only matches consecutive elements. this is why I have the variant expression with the wildcard, to match things like comments or `"::"` nodes.

**Comment by MixusMinimax** [comment]
Intuitively, it would make sense if this expression matches subsets of a list of identifiers, but that is not the case for the CLI, which is why I was confused about the web-tree-sitter case.

Maybe using matches instead of captures is the right thing to do anyway? Now that I think about it in that way, that sounds reasonable.

**Reply by ahlinc** [reply]
Could you give a link to the grammar's repo if it's possible?

**Reply by MixusMinimax** [reply]
I have set up a minimal example here: [gh:MixusMinimax/tree-sitter-test](https://github.com/MixusMinimax/tree-sitter-test)

You can see the relevant output in the 5th step "Run npm run-script test-node".

Edit: it seems like other people can't see the action logs. I will put the output here:

<details>

<summary><code>npm run-script test-node</code></summary>
<br/>

```js
Running example example1

Code:
module core::event;

Tree:
(source_file (module path: (path (identifier) (identifier))))

Captures:
{ index: 0, name: 'namespace', nodeId: 92872, text: 'core' }
{ index: 1, name: 'namespace', nodeId: 92872, text: 'core' }
{ index: 2, name: 'namespace', nodeId: 92872, text: 'core' }
{ index: 3, name: 'namespace', nodeId: 92616, text: 'event' }
{ index: 4, name: 'namespace', nodeId: 92616, text: 'event' }
{ index: 5, name: 'namespace', nodeId: 92616, text: 'event' }

Matches:
{
  index: 0,
  pattern: 0,
  captures: [
    { index: 0, name: 'namespace', nodeId: 92872, text: 'core' },
    { index: 1, name: 'namespace', nodeId: 92616, text: 'event' }
  ]
}
```
</details>

**Comment by ahlinc** [comment, marked as answer]
> The result using the rust cli
> Result using web-tree-sitter

Don't mix _captures_ and _matches_ modes, that was the main mistake.

Your original issue reproducible with CLI so it's not a bug in the web binding:

![image](https://github.com/tree-sitter/tree-sitter/assets/14666676/7345281b-eb26-4057-ba19-97033a02c84d)

In the _matches_ mode your original query works as you expected:

![image](https://github.com/tree-sitter/tree-sitter/assets/14666676/32d4d6b4-7ab7-4de9-be9c-de6b291bc3f0)

You need to properly construct your queries, on the bellow screenshot there are fixed queries for both _captures_ and _matches_ modes.  Keep in mind that queries for the _matches_ mode may not work well for the _captures_ mode. Also be attentive when you use underscore placeholders, usually it's better to use explicit node names and especially for string literal terminals.

![image](https://github.com/tree-sitter/tree-sitter/assets/14666676/b7e0e5e3-0ceb-4cda-936d-661f75797cda)

**Reply by MixusMinimax** [reply]
The reason why I used the variant and not your last example, is that if I for example have this string:

```
module core /**/   /**/  ::     /**/ // asdasd
events;
```

That query is not robust against extras in between nodes. I need to be able to query a list of identifiers, no matter how many other nodes are in between them.

But I think I understand how captures work a little better now, I guess just using matches instead is better.

But the captures query (where you do not use a quantifier) Works well. The only issue with that would be that it doesn't detect the difference between multiple module declarations or one with multiple path segments.

So I will use my original query, but not use the global captures list from now on. Thanks for your help!

My mistake was that I thought the CLI used captures per default, but it is of course now obvious to me that it is using matches mode.

**Reply by unennhexium** [reply]
@ahlinc, what is the difference between the modes? I have duplicates with both: captures `tree-sitter q -c` and matches `tree-sitter q`. I have the query similar to `a[bc]*(d[bc]*(e[bc]*f?[bc]*)+)+` regex. I am querying markdown: `a`, `d` and `e` are document, section and subsection headers; `[bc]` is either paragraph or fenced code-block; `f` is the list. I am @-tagging each of these nodes.

**Reply by sogaiu** [reply]
AFAIU, unfortunately, ahlinc no longer participates.

**Reply by unennhexium** [reply]
@sogaiu, [here] @maxbrunsfeld claimed it is intended behavior. For what mode? It means that for every end (deepest) match I will have the entire chain of parents, hence duplicates. But for which mode this claim is in affect. Matches? Captures?

[here]: https://github.com/tree-sitter/tree-sitter/discussions/2101#discussioncomment-5105323

**Reply by sogaiu** [reply]
> what is the difference between the modes? 

I don't know, but the following links suggest output order may differ:

* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/docs/src/cli/query.md#-c--captures
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/main.rs#L448-L450
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/main.rs#L1426-L1431
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/main.rs#L1465-L1470
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/main.rs#L1487-L1492
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/query.rs#L21-L26
* https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/query.rs#L63-L117

**Reply by unennhexium** [reply]
@sogaiu, I have understand what does go wrong with my query. See the [explanation](https://github.com/tree-sitter/tree-sitter/discussions/4977#discussioncomment-14732887). Thanks for the discussion. 😺

**Reply by unennhexium** [reply]
@sogaiu, I am sorry, I understand you wrong. I checked the repo and your link more thoroughly and realized the concept.

```
tree-sitter C library -> C API
                      -> Rust bindings -> CLI
                      -> WASM bindings aka web-tree-sitter
```

So you refer to the CLI code, which is the Rust code. The C library contains corresponding functions.

- Advance in order of captures
  - https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/query.rs#L64
  - https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/lib/src/query.c#L4250

- Advance in order of matches (more complicated)
  - https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/crates/cli/src/query.rs#L86
  - https://github.com/tree-sitter/tree-sitter/blob/da5926d6f571475625bd631c593b255c437e9135/lib/src/query.c#L3637
